library/src/main/java/org/apache/apex/malhar/lib/fs/s3output/S3BlockUpload.java

```diff
@@ -0,0 +1,516 @@
+import com.datatorrent.api.DefaultOutputPort;
+import com.datatorrent.api.Operator;
+import com.datatorrent.lib.io.block.AbstractBlockReader;
+import com.datatorrent.lib.io.block.BlockMetadata;
+import com.datatorrent.netlet.util.Slice;
+
+/**
+ * This operator can be used to upload the block into S3 bucket using multi-part feature or putObject API.
+ * Upload the block into S3 using multi-part feature only if the number of blocks of a file is > 1.
+ */
+
+@InterfaceStability.Evolving
+public class S3BlockUpload implements Operator, Operator.CheckpointNotificationListener
+{
+  private static final Logger LOG = LoggerFactory.getLogger(S3BlockUpload.class);
+  @NotNull
+  private String bucketName;
+  @NotNull
+  private String accessKey;
+  @NotNull
+  private String secretAccessKey;
+  private String endPoint;
+  private Map<String, S3BlockMetaData> blockInfo = new HashMap<>();
+  private Map<Long, String> blockId2FilePath = new HashMap<>();
+  private WindowDataManager windowDataManager = new FSWindowDataManager();
+  protected transient AmazonS3 s3Client;
+  private transient long currentWindowId;
+  private transient List<AbstractBlockReader.ReaderRecord<Slice>> waitingTuples;
+  private transient Map<String, UploadBlockMetadata> currentWindowRecoveryState;
+  public final transient DefaultOutputPort<UploadBlockMetadata> output = new DefaultOutputPort<>();
+
+  /**
+   * This input port receives incoming tuples.
+   */
+  public final transient DefaultInputPort<AbstractBlockReader.ReaderRecord<Slice>> blockInput = new DefaultInputPort<AbstractBlockReader.ReaderRecord<Slice>>()
+  {
+    @Override
+    public void process(AbstractBlockReader.ReaderRecord<Slice> tuple)
+    {
+      uploadBlockIntoS3(tuple);
+    }
+  };
+
+  /**
+   * Input port to receive Block meta data
+   */
+  public final transient DefaultInputPort<BlockMetadata.FileBlockMetadata> blockMetadataInput = new DefaultInputPort<BlockMetadata.FileBlockMetadata>()
+  {
+    @Override
+    public void process(BlockMetadata.FileBlockMetadata blockMetadata)
+    {
+      if (currentWindowId <= windowDataManager.getLargestCompletedWindow()) {
+        return;
+      }
+      blockId2FilePath.put(blockMetadata.getBlockId(), blockMetadata.getFilePath());
+      LOG.debug("received blockId {} for file {} ", blockMetadata.getBlockId(), blockMetadata.getFilePath());
+    }
+  };
+
+  /**
+   * Input port to receive upload file meta data.
+   */
+  public final transient DefaultInputPort<S3InitiateFileUpload.UploadFileMetadata> uploadMetadataInput = new DefaultInputPort<S3InitiateFileUpload.UploadFileMetadata>()
+  {
+    @Override
+    public void process(S3InitiateFileUpload.UploadFileMetadata tuple)
+    {
+      processUploadFileMetadata(tuple);
+    }
+  };
+
+  /**
+   * Convert each block of a given file into S3BlockMetaData
+   * @param tuple UploadFileMetadata
+   */
+  protected void processUploadFileMetadata(S3InitiateFileUpload.UploadFileMetadata tuple)
+  {
```

Comment:
This constructs the blockInfo Map and clearing this in replay call. So, I think it's better to process even if currentWindowId <= windowDataManager.getLargestCompletedWindow().